clients/src/test/java/org/apache/kafka/clients/NetworkClientApiVersionsCheckTest.java

```diff
@@ -0,0 +1,102 @@
+import static org.apache.kafka.common.requests.ApiVersionsResponse.MIN_VERSION_KEY_NAME;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+public class NetworkClientApiVersionsCheckTest extends NetworkClientTest {
+
+    private static final ApiKeys USED_API_KEY = ApiKeys.METADATA;
+    private static final Collection<ApiVersionsResponse.ApiVersion> EXPECTED_API_VERSIONS = createExpectedApiVersions();
+
+    private static Collection<ApiVersionsResponse.ApiVersion> createExpectedApiVersions() {
+        List<ApiVersionsResponse.ApiVersion> expectedApiVersions = new ArrayList<>();
+        expectedApiVersions.add(new ApiVersionsResponse.ApiVersion(USED_API_KEY.id, Protocol.MIN_VERSIONS[USED_API_KEY.id], Protocol.CURR_VERSION[USED_API_KEY.id]));
+        return expectedApiVersions;
+    }
+
+    @Override
+    protected Collection<ApiVersionsResponse.ApiVersion> getExpectedApiVersions() {
+        return EXPECTED_API_VERSIONS;
+    }
+
+    @Test
+    public void testUnsupportedLesserApiVersions() {
+        unsupportedApiVersionsCheck(getExpectedApiVersions(), (short) (ProtoUtils.latestVersion(USED_API_KEY.id) + 1), Short.MAX_VALUE, "Node 0 does not support required versions for Api " + USED_API_KEY.id);
+    }
+
+    @Test
+    public void testUnsupportedGreaterApiVersions() {
+        unsupportedApiVersionsCheck(getExpectedApiVersions(), Short.MIN_VALUE, (short) (ProtoUtils.oldestVersion(USED_API_KEY.id) - 1), "Node 0 does not support required versions for Api " + USED_API_KEY.id);
+    }
+
+    @Test
+    public void testUnsupportedMissingApiVersions() {
+        unsupportedApiVersionsCheck(new ArrayList<ApiVersionsResponse.ApiVersion>(), Short.MIN_VALUE, (short) (Short.MAX_VALUE + 2), "Node 0 does not support Api " + USED_API_KEY.id);
+    }
+
+    private void unsupportedApiVersionsCheck(final Collection<ApiVersionsResponse.ApiVersion> expectedApiVersions, short minVersion, short maxVersion, String errorMessage) {
+        ResponseHeader respHeader = new ResponseHeader(0);
+        Struct resp = new Struct(ProtoUtils.currentResponseSchema(ApiKeys.API_VERSIONS.id));
+        resp.set(ERROR_CODE_KEY_NAME, (short) 0);
+        List<Struct> apiVersionList = new ArrayList<>();
+        for (ApiVersionsResponse.ApiVersion apiVersion : expectedApiVersions) {
+            Struct apiVersionStruct = resp.instance(API_VERSIONS_KEY_NAME);
+            apiVersionStruct.set(API_KEY_NAME, apiVersion.apiKey);
+            apiVersionStruct.set(MIN_VERSION_KEY_NAME, minVersion);
+            apiVersionStruct.set(MAX_VERSION_KEY_NAME, maxVersion);
+            apiVersionList.add(apiVersionStruct);
+        }
+        resp.set(API_VERSIONS_KEY_NAME, apiVersionList.toArray());
+        int size = respHeader.sizeOf() + resp.sizeOf();
+        ByteBuffer buffer = ByteBuffer.allocate(size);
+        respHeader.writeTo(buffer);
+        resp.writeTo(buffer);
```

Comment:
Don't we have some code that we can reuse for this? Why are we manually serialising `ApiVersionResponse`?